### PR TITLE
PP-5391 centralise refund state transition

### PR DESCRIPTION
**Pull request by SandorArpa**
## WHAT YOU DID
- refactor setting the status of refund entity in ChargeRefundService.
This work is to lay the grounds for emitting refund events to ledger from a single point.

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/gateway/processor/RefundNotificationProcessor.java [hunk 1]
```diff
@@ -17,13 +17,13 @@
 public class RefundNotificationProcessor {
 
     private Logger logger = LoggerFactory.getLogger(getClass());
-    private RefundDao refundDao;
+    private ChargeRefundService refundService;
```
Comment (sfount) [inline src/main/java/uk/gov/pay/connector/gateway/processor/RefundNotificationProcessor.java:20]: 👍 using the higher level service in favour of many things accessing the DAO

File: src/test/java/uk/gov/pay/connector/model/domain/RefundEntityTest.java [hunk 2]
```diff
@@ -26,12 +26,13 @@ public void shouldConstructANewEntity() {
...
 
         RefundEntity refundEntity = new RefundEntity(chargeEntity, amount, userExternalId);
+        refundEntity.setStatus(CREATED);
 
         assertNotNull(refundEntity.getExternalId());
         assertThat(refundEntity.getChargeEntity(), is(chargeEntity));
         assertThat(refundEntity.getReference(), is(nullValue()));
         assertThat(refundEntity.getAmount(), is(amount));
-        assertThat(refundEntity.getStatus(), is(RefundStatus.CREATED));
+        assertThat(refundEntity.getStatus(), is(CREATED));
```
Comment (sfount) [inline src/test/java/uk/gov/pay/connector/model/domain/RefundEntityTest.java]: is there any benefit to asserting this is created after having set it in the test itself? Potentially this assertion should be made on a test for the `createRefundEntity` which both created the entity as transitions the status to `CREATED`

File: src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java [hunk 3]
```diff
@@ -163,12 +167,17 @@ private RefundStatus determineRefundStatus(GatewayRefundResponse gatewayRefundRe
...
     public RefundEntity createRefundEntity(RefundRequest refundRequest, ChargeEntity charge) {
         RefundEntity refundEntity = new RefundEntity(charge, refundRequest.getAmount(), refundRequest.getUserExternalId());
+        transitionRefundState(refundEntity, RefundStatus.CREATED);
         charge.getRefunds().add(refundEntity);
         refundDao.persist(refundEntity);
 
         return refundEntity;
     }
 
+    public void transitionRefundState(RefundEntity refundEntity, RefundStatus refundStatus) {
```
Comment (sfount) [inline src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java:177]: This should probably set `@transactional` to make sure it's getting the latest version of everything. I'm not 100% with JPA but ran into some issues without this annotation in Connector

**Review by sfount — COMMENTED** [review]
Looks good! Few comments about testing

File: src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java (hunk 3 above)
Comment (SandorArpa) [reply src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java:177]: All the methods that this one is called from are marked @Transactional. If we mark this method as well there will be 2 entry in refund events for the given nature of how state change (or any other commits to db) handled (see `RefundHistory`)

File: src/test/java/uk/gov/pay/connector/model/domain/RefundEntityTest.java (hunk 2 above)
Comment (SandorArpa) [reply src/test/java/uk/gov/pay/connector/model/domain/RefundEntityTest.java]: Added a test in `ChargeRefundServiceTest` for `createRefundEntity`

File: src/test/java/uk/gov/pay/connector/service/ChargeRefundServiceTest.java [hunk 4]
```diff
@@ -137,6 +137,32 @@ public void shouldRefundSuccessfully_forWorldpay() {
...
+                .build();
+
+        doAnswer(invocation -> {
+            ((RefundEntity) invocation.getArgument(0)).setId(refundId);
+            return null;
+        }).when(mockRefundDao).persist(any(RefundEntity.class));
+        RefundEntity refundEntity = chargeRefundService.createRefundEntity(new RefundRequest(refundAmount, charge.getAmount(), userExternalId), charge);
+
+        assertThat(refundEntity.getAmount(), is(refundAmount));
+        assertThat(refundEntity.getStatus(), is(RefundStatus.CREATED));
```
Comment (sfount) [inline src/test/java/uk/gov/pay/connector/service/ChargeRefundServiceTest.java:163]: 👍

**Review by sfount — APPROVED** [review]

**Pull request merged**